### DISCUSSION TITLE
Mesh slicing

### DISCUSSION THREAD

**Discussion by MarcusLoppe** [lucidrains/meshgpt-pytorch#25] [Ideas]
I might have misunderstood some parts of the encoding & decoding so correct me if I'm wrong.

Current the SageConv takes all the faces in a F x init_dim size, and outputs F x 576 , when the face count gets high enough it will run out of memory very fast. Since the encoder needs all the data for all the other faces it cannot do this in batches.
The decoder have the same type of issues.

The current state of doing things I don't foresee any high face count meshes being generated or trained on even if there is a x10 improvements. The issue remains; the entire mesh needs to be processed at once.

I got a kinda unfinished idea but since it's a discussion page I was hoping to discuss it :)

Instead of processing all at once, you use a fixed width input for the encoder & decoder.
 
So lets say it needs it inputs the faces that is marked in yellow, you take these face features and input it into a fixed padded space.
Then for extra information about the rest of the shape you take the face features from the nearest edges of the faces that needs processing (green) and input as well in a fixed padded space.

With the rest of the faces you create a embedding e.g face feature summarizer. If the shape is a ball then it might not need the full mesh to get a understanding of the shape but a embedding that describe the shape.
 

Hopefully this will let you scale to 15k poly count and beyond. Since people want these high poly counts and not simple shapes; I think slicing it is the only way. 

Here is a beautiful image that will try to describe it.
![bild](https://github.com/lucidrains/meshgpt-pytorch/assets/65302107/b7f34c5c-253d-478f-a2b3-82af9d8b7ccf)

**Comment by lucidrains** [comment]
if you get something to work, you can write a follow up paper

something something hierarchical embedding

**Reply by MarcusLoppe** [reply]
That seems unlikely since I can't write & don't fully understand the current code implementation :) 

The current SageConv setup seems to do pretty well, I don't think it's much overhead to embed 14k triangles but the issue  is when the ResNet decodes it. 

Here's some preliminary results of testing with the SageConv sizes; I used 530 3d meshes from many categories from the shapenet dataset, I applied 10 augmentations per mesh and duplicated each augment. So 530 * 10 * 2 = 10 600 dataset.

After 9hrs of training on two kaggle notebooks, the smaller 661k  SageConv got 0.57 loss while a bigger 7 M got 0.66. I'm also testing using the old dim x 6 setup but that is running atm.

The tests shows that the 7 M network was just a 2 second slowdown on a 10 600 dataset vs a 661k network. 
This means that the embedding is very light weight and main slowdown is the ResNet.
Using SageConv to embed 14k faces doesn't seem that it would impact the performance much. Instead of letting the ResNet process all the faces at once and guess, why let it process it 100 faces + embedding of the rest of the triangles? This would speed up the training + inference. 
And adding the neighbouring faces without letting it predict on them will add some "ground truth".

Please critic me :) Let me know if I've misunderstand anything so I'm not walking in the wrong direction.

**SageConv 661k:**
```
Total parameters: 661 376
ModuleList(
  (0): SAGEConv(64, 128, aggr=mean)
  (1): SAGEConv(128, 256, aggr=mean)
  (2): SAGEConv(256, 256, aggr=mean)
  (3): SAGEConv(256, 576, aggr=mean)
)
Epoch 180/180: 100%|██████████| 441/441 [02:48<00:00,  loss=0.6]  
Epoch 180 average loss: 0.5797933419305392
Training complete
```
 
**SageConv  7M:**
```
Encoders Total parameters: 7052608
ModuleList(
  (0): SAGEConv(256, 512, aggr=mean)
  (1): SAGEConv(512, 1024, aggr=mean)
  (2): SAGEConv(1024, 512, aggr=mean)
  (3): SAGEConv(512, 512, aggr=mean)
  (4): SAGEConv(512, 768, aggr=mean)
  (5): SAGEConv(768, 576, aggr=mean)
)
Epoch 180/180: 100%|██████████| 441/441 [02:50<00:00,   loss=0.708]
Epoch 180 average loss: 0.6670283106178924
```

**Reply by lucidrains** [reply]
i think you'll get there within a year. what you are doing, poking around, changing layers, generating loss curves and evaluated results etc, nothing different than what a graduate student does

![data](https://github.com/lucidrains/meshgpt-pytorch/assets/108653/0c72e327-b771-4ba9-8662-58aabf63769e)

**Reply by MarcusLoppe** [reply]
Okay but what do you think about the idea?
Do you have another way to reducing the memory requirement?
The issue is that it takes the entire mesh while predicting, if it needs to process 14000x576 and output 14000x128x3 it will take up a huge amount of memory, do you have any other idea then to slice it?


"We reach out and reach out and flippin' and flippin' the switches like a goddamn monkey to see if something clicks."

**Reply by lucidrains** [reply]
@MarcusLoppe yea i think the general idea is sound

instead of one global embedding, i would simply train another mesh network that handles a low resolution mesh. then perhaps some way to super resolute local areas with an encoder-decoder

**Reply by lucidrains** [reply]
@MarcusLoppe i'm not in this field, so you'd probably want to discuss this with a phd student who is actually working in this niche

**Reply by lucidrains** [reply]
you can probably do a literature review with keywords like `3d, mesh, hierarchical` for inspiration

**Reply by MarcusLoppe** [reply]
> @MarcusLoppe yea i think the general idea is sound
> 
> instead of one global embedding, i would simply train another mesh network that handles a low resolution mesh. then perhaps some way to super resolute local areas with an encoder-decoder

Not sure if that would fix anything since if the decoder just decode the positions with less accuracy. 
That would require that low-resolution mesh be also able to process the entire mesh, hence why a embedding might be better.

> @MarcusLoppe i'm not in this field, so you'd probably want to discuss this with a phd student who is actually working in this niche

I'll check with my neighbours :) Nah but I might be able to reach out to the person whom published the MeshGPT paper. They probably tested with many different things before using the current setup.

**Reply by MarcusLoppe** [reply]
> you can probably do a literature review with keywords like `3d, mesh, hierarchical` for inspiration

I'll check it out, but meanwhile. Do you have any idea on how to scale to higher poly count meshes?

**Reply by lucidrains** [reply]
you are at the forefront of research, so nobody knows the answer. you'd be very famous if you find a good solution for it though!

**Reply by MarcusLoppe** [reply]
oh to become a **rich** and famous research <3

One another idea is to use quads instead of triangles, a quad just contains one more vertex and might reduce the shape count to 50%

**Reply by MarcusLoppe** [reply]
> you are at the forefront of research, so nobody knows the answer. you'd be very famous if you find a good solution for it though!

But what is your plan for this repo? I mean is it meant only to generate max 800-1.3k face meshes? 
That isn't very useful if someone want a complex model :/

**Reply by lucidrains** [reply]
my plan is to simply open source new SOTA model architectures

I think that has been achieved

**Reply by lucidrains** [reply]
just be patient, if the paper does not have flaws, there may be a Robin Rombach in store for 3d meshes. may be just a few more papers away

**Reply by MarcusLoppe** [reply]
> my plan is to simply open source new SOTA model architectures
> 
> I think that has been achieved

Fair enough, but why not now? :) 

But what do you think leads to the high VRAM usage? It is simple feeding it a vector that is multiplied by F? 
If so some ideas:

- 8/16 bit - Quantize or support natively.
- Replace the ResNet with a much more effective network, ResNet perhaps isn't SOTA.
- Testing required; but if the transformer network itself is slow,Fast Feedforward Network might can help. 


edit:
Yinks, I get now why they needed to train for 4 days. Avg epoch time for encoder; 5 min, estimated for the transformer: 7 hours!
😮 

Autoencoder: Epoch 180/180: 100%|██████████| 2650/2650 [05:47<00:00,  7.62it/s, loss=0.496]
Transformer: Epoch 1/10:   7%|▋         | 720/10600 [16:21<6:57:42,  2.54s/it, loss=8.34]

**Reply by lucidrains** [reply]
@MarcusLoppe the reason is because i don't see myself being able to open source indefinitely and there are other projects i need to make headway on

**Reply by lucidrains** [reply]
in research, you can spend a lot of time trying to come up with some solution, only to come up empty handed half a year later.

**Reply by lucidrains** [reply]
that said, hope the repository empowers you (and others) to make the next improvement

**Reply by MarcusLoppe** [reply]
@lucidrains  I understand, the curve for improvements isn't very linear and dead-ends are far too common.

Just by estimating,  how much work would it require to instead of feeding the entire mesh that you limit it to 100 faces + a embedding of the rest of the faces that is lets say, 200x576?

So there would be two SageConv networks, one with that creates a fixed embedding for the rest of the faces plus the one with the faces the ResNet needs to predict.
So the input for the ResNet would be:  [ Overview_Embedding(200x576) +  F x 576 ]

**Reply by lucidrains** [reply]
@MarcusLoppe it cannot be estimated as this is still an open research question, afaik
